### PR TITLE
Add masking of data with line quality flags to seviri nc reader

### LINKED ISSUE #1482: Add option to turn off masking in seviri_l1b_hrit
## Feature Request

**Is your feature request related to a problem? Please describe.**
Currently you cannot disable the masking of bad quality scanlines in `seviri_l1b_hrit`.

**Describe the solution you'd like**
Add an option to turn the masking off.

**Describe any changes to existing user workflow**
No.

### PR DESCRIPTION
**Pull request by BENR0**
This PR adds masking of data with the line quality flags to the Seviri nc reader. In the Seviri hrit reader this masking is already implemented and I think this behavior should be consistent between the readers.

I refactored the main logic of the masking out of the hrit reader to reuse it in the nc reader. There is no test for the masking itself yet (by the way I noticed that it isn't tested in the hrit reader either). Will add the test if this get's a go.

Just saw #1482 which is related and would be nice to have.

 - [x] Closes #1482 
 - [x] Tests added 
 - [x] Fully documented

### CODE REVIEW

File: satpy/readers/seviri_base.py
Comment (sfinkens) [inline satpy/readers/seviri_base.py:915]: To improve readability this method could be split into smaller methods like so

```
def mask_bad_quality(data, line_flags):
    line_mask = _get_bad_quality_lines(*line_flags)
    return _mask_bad_quality_lines(data, line_mask)
```

Even if `_mask_bad_quality_lines` is a one-liner, it helps keeping the contents of `mask_bad_quality` at the same level of abstraction.

File: satpy/readers/seviri_base.py
Comment (sfinkens) [inline satpy/readers/seviri_base.py:942]: I think it would be clearer to create a masked copy of the data instead of modifying it inplace. What do you think?

File: satpy/readers/seviri_l1b_hrit.py
Comment (sfinkens) [inline satpy/readers/seviri_l1b_hrit.py:583]: You can delete the comments here now

File: satpy/tests/reader_tests/test_seviri_l1b_nc.py
Comment (sfinkens) [inline satpy/tests/reader_tests/test_seviri_l1b_nc.py:191]: Is it possible to store the `ch1/9` keys in a global dictionary and look them up inside the test function? That would keep the parametrization simpler.

**Review by sfinkens — CHANGES_REQUESTED** [review]
Nice work, thanks a lot! Just a few comments inspired by Uncle Bob.

If you have time, it would indeed be nice to make the masking optional. Otherwise, we can leave it to another PR.

File: satpy/readers/seviri_base.py
Comment (BENR0) [reply satpy/readers/seviri_base.py:915]: I am not sure if this would improved readability since it fragments the code into really small pieces. I mean the additional functions that would be created would have no real functional purpose on their own. But if this is wanted I can refactor.

File: satpy/readers/seviri_base.py
Comment (BENR0) [reply satpy/readers/seviri_base.py:942]: Maybe yes. I thought about that too but wanted to keep it as close as possible to the original code for the review.

File: satpy/readers/seviri_base.py
Comment (sfinkens) [reply satpy/readers/seviri_base.py:915]: Yes, this leads to many small pieces. But if named properly they document the behavior very well. You can read the code like a news article from top to bottom with increasing level of detail. It should be possible for readers to quickly grasp what the method is doing and then decide whether they want to know the details.

Note that I'm just reading the "Clean Code" book by Robert C. Martin and are very enthusiastic  about it :)

**Comment by BENR0** [comment]
@sfinkens @djhoese @mraspaud @adybbroe what's the status on this? Is this still wanted? I am still not convinced on the refactoring suggested by @sfinkens since it adds unnecessary complexity from my point of view.

File: satpy/readers/seviri_base.py
Comment (djhoese) [reply satpy/readers/seviri_base.py:915]: I'm not against what @sfinkens has said, but I don't completely agree either. The result he posted in his original comment is indeed easier to read, but I'm not sure it is necessary based on the level of abstraction/operations being done in the function. Yes, what he suggested would be considered best practice as far as "Clean Code" goes, but this might be a case where it isn't a big deal to disobey it. I could also see an argument for combining the line_mask lines into one or two lines (combining the `&`s) or renaming this method to `mask_bad_quality_lines` so it is more clear it is working with lines. The last thing is that the `np.choose` line is ugly (because it has to be) and with a lot of arguments that aren't arguments to the function (np.nan, np.newaxis, etc) so it feels like it should be refactored.

My final opinion is to not refactor it. Leave it as is. But I'd like to hear other opinions.

**Comment by BENR0** [comment]
So it seems there are no other opinions. So I guess it's ready to be merged ;-).

**Comment by mraspaud** [comment]
Just some conflicts with main to solve...

**Comment by BENR0** [comment]
@mraspaud @djhoese I think this is ready to go then.

**Comment by simonrp84** [comment]
Just to check I'm reading the code correctly: Am I right in thinking that this masking is optional and is disabled by default?

**Comment by BENR0** [comment]
@simonrp84 no this is implemented the same as in the hrit reader. I think the only way to not mask the data is to request counts:
https://github.com/pytroll/satpy/blob/de969f554f3f69a5561c9159868979e27489c680/satpy/readers/seviri_l1b_nc.py#L140
https://github.com/pytroll/satpy/blob/de969f554f3f69a5561c9159868979e27489c680/satpy/readers/seviri_l1b_hrit.py#L680

**Comment by djhoese** [comment]
pre-commit.ci autofix

**Comment by simonrp84** [comment]
I think it's important to have the masking as optional (vaguely remember this discussion in the past with the HRIT reader). But as both HRIT and NC don't do that then it's something for another issue / PR.

File: satpy/readers/seviri_base.py
Comment (mraspaud) [reply satpy/readers/seviri_base.py:928]: I'm with @sfinkens here.

File: satpy/readers/seviri_base.py
Comment (mraspaud) [reply satpy/readers/seviri_base.py:953]: Thank you for thinking of the reviewers ! :smile: 
However, if you see something that can be refactored to be simpler, you don't need to hesitate, just do it!

My suggestion here would be
```suggestion
    line_mask = line_mask[:, np.newaxis]
    data = np.where(line_mask, data, np.nan).astype(np.float32)
```

**Comment by BENR0** [comment]
@djhoese @mraspaud can this also be merged?

**Comment by mraspaud** [comment]
I think there are a couple of unresolved issue left @BENR0 , right?

**Comment by djhoese** [comment]
Looks like @sfinkens had some style concerns that he wanted fixed and @simonrp84 had some concerns about default behavior versus a kwarg.

**Comment by BENR0** [comment]
Since there were no other opinions raised after https://github.com/pytroll/satpy/pull/1693#discussion_r871303238 my take was that this is good to go code wise. I agree to leave the concern regarding the optional masking raised by @simonrp84 for another PR. This PR just makes sure that the hrit and nc reader do the same processing which I think is important from a user perspective.

**Comment by djhoese** [comment]
Sounds good with the Simon comment. The refactor comment: I said don't refactor it, but Martin agreed with Stephan so I got overruled. I think they would both like if you refactored that bit of code before they're OK with merging this.

File: satpy/readers/seviri_base.py
Comment (BENR0) [reply satpy/readers/seviri_base.py]: yep. I changed that.

File: satpy/readers/seviri_base.py
Comment (BENR0) [reply satpy/readers/seviri_base.py]: I refactored the creation of the scan lines mask into a separate function. Hope this is how you wanted it to look like.

**Comment by BENR0** [comment]
pre-commit.ci autofix

**Comment by BENR0** [comment]
> Just to check I'm reading the code correctly: Am I right in thinking that this masking is optional and is disabled by default?

@simonrp84 I also added the masking to be optional now.

**Review by sfinkens — CHANGES_REQUESTED** [review]
Almost there! Thanks for adding more calibration tests for the netCDF reader.

File: satpy/tests/reader_tests/test_seviri_l1b_hrit.py
Comment (BENR0) [reply satpy/tests/reader_tests/test_seviri_l1b_hrit.py]: Done. Actually these parametrizations weren't really necessary at all in this test so I removed them completely.

File: satpy/tests/reader_tests/test_seviri_l1b_hrit.py
Comment (BENR0) [reply satpy/tests/reader_tests/test_seviri_l1b_hrit.py:498]: Not really. To adhere to unit test best practice this test is only testing this method explicitly (like the test for the calibration above), which I can not do if I use the `get_dataset` method because there are other things happening apart from the masking. If I remember correctly the method was public at first like the `calibrate` method but I think I made this method private based on a wish from @mraspaud. I agree though that it makes sense that it is private. Anyway in this case the "tight coupling" is not a big problem I suppose.

Comment (sfinkens) [reply satpy/tests/reader_tests/test_seviri_l1b_hrit.py:498]: Ok, fine by me :)

File: satpy/tests/reader_tests/test_seviri_l1b_nc.py
Comment (sfinkens) [inline satpy/tests/reader_tests/test_seviri_l1b_nc.py:336]: I know I'm being a bit picky, but you're using the method under test to compute the expected values ;) Can you define them explicitly?

**Review by sfinkens — CHANGES_REQUESTED** [review]
Just nitpicking, otherwise looking good now!

File: satpy/readers/seviri_l1b_hrit.py
Comment (BENR0) [reply satpy/readers/seviri_l1b_hrit.py]: I totally agree that this is not really nice and readability is somewhat impaired. I have a feeling that this change request is also inspired by the "Clean Code" book ;-). In this case I have to strongly disagree though that refactoring this into a separate method would improve readability. This would fragment the code a lot with yet another private method which has no real purpose. While it might improve the one line ,the "unattractive" code would just be moved somewhere else and a reader would need to jump to a method with no body just to have a look at what conditions are applied here. My proposition (which is included in the last push) is to move the list check to a separate line which improves readability and keeps the code lean an close together.

File: satpy/tests/reader_tests/test_seviri_l1b_nc.py
Comment (BENR0) [reply satpy/tests/reader_tests/test_seviri_l1b_nc.py:336]: I probably could. In a discussion with @mraspaud at PCW we found that it is ok to use it here because `_mask_bad_quality` is tested in a separate unit test on its own. Thus if the test passes it can safely be used here.

File: satpy/readers/seviri_l1b_hrit.py
Comment (sfinkens) [reply satpy/readers/seviri_l1b_hrit.py]: I disagree (maybe I've watched too many Uncle Bob videos), but your proposed solution sounds fine! :)

File: satpy/tests/reader_tests/test_seviri_l1b_nc.py
Comment (sfinkens) [reply satpy/tests/reader_tests/test_seviri_l1b_nc.py:336]: Ok, sounds good!

**Review by sfinkens — APPROVED** [review]
LGTM, thanks for all the hard work!

**Pull request merged**
